// exiftool-vendored/dist/Object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isObject = isObject;
exports.keys = keys;
exports.isFunction = isFunction;
exports.fromEntries = fromEntries;
exports.omit = omit;
exports.keysOf = keysOf;
function isObject(obj) {
    return typeof obj === "object" && obj !== null;
}
function keys(o) {
    return o == null
        ? []
        : Object.keys(o).filter((ea) => ({}).propertyIsEnumerable.call(o, ea));
}
function isFunction(obj) {
    return typeof obj === "function";
}
function fromEntries(arr, obj) {
    if (arr == null || arr.length === 0)
        return obj ?? {};
    // don't use Object.create(null), json stringify will break!
    for (const ea of arr.filter((ea) => ea != null)) {
        if (ea != null && Array.isArray(ea)) {
            const [k, v] = ea;
            // allow NULL fields:
            if (k != null && v !== undefined) {
                if (!isObject(obj))
                    obj = {};
                obj[k] = v;
            }
        }
    }
    return obj ?? {};
}
function omit(t, ...keysToOmit) {
    const result = {};
    for (const k of keys(t).filter((ea) => !keysToOmit.includes(ea))) {
        result[k] = t[k];
    }
    return result;
}
/**
 * Provides a type-safe exhaustive array of keys for a given interface.
 *
 * Unfortunately, `satisfies (keyof T)[]` doesn't ensure all keys are present,
 * and doesn't guard against duplicates. This function does.
 *
 * @param t - The interface to extract keys from. This is a Record of keys to
 * `true`, which ensures the returned key array is unique.
 */
function keysOf(t) {
    return Object.keys(t);
}
// This also doesn't enforce that all keys are present:
// type RequiredKeys<T> = { [K in keyof Required<T>]: K } extends { [K: string]: infer U } ? U[] : never;
